src/main/java/uk/gov/hmcts/reform/iahearingsapi/domain/mappers/LanguageAndAdjustmentsMapper.java

```diff
@@ -0,0 +1,171 @@
+    public static final Set<String> REASONABLE_ADJUSTMENT_PREFIXES = Set.of("RA", "SM");
+
+    public static final String INTERPRETER_LANGUAGE = "interpreterLanguage";
+    public static final String OTHER_REASONABLE_ADJUSTMENTS_DETAILS = "otherReasonableAdjustmentsDetails";
+    public static final String REASONABLE_ADJUSTMENTS = "reasonableAdjustments";
+
+    public static final String INTERPRETER = "Interpreter: ";
+
+    /**
+     * Method to extract the interpreter language, other reasonable adjustments details and reasonable adjustments.
+     * These three fields are extrapolated by this same method, as they don't relate to the three hearing values in
+     * a 1:1 way.
+     * Interpreter Language or Sign Language Interpreter flags: the first one with code (e.g. "deu") populates
+     * the "interpreterLanguage" value, any additional one with code will populate the
+     * "otherReasonableAdjustmentsDetails" value, any manually typed one (without the "code") will add to
+     * "otherReasonableAdjustmentsDetails" too.
+     * Reasonable Adjustment flags (with flagCode "RA****" or "SM****") except Sign Language Interpreter: the codes
+     * of each will populate the "reasonableAdjustments" value, the comment of each, if present, will add to the
+     * "otherReasonableAdjustmentsDetails" value.
+     * @param asylumCase The asylum case queried for a hearing.
+     * @return           A map with the three values to be written in the response to HMC.
+     */
+    public Map<String, List<String>> getLanguageAndAdjustmentsFields(AsylumCase asylumCase) {
+        Map<String, List<String>> fields = new HashMap<>();
+
+        List<CaseFlagDetail> caseFlags = getAppellantAndWitnessCaseFlags(asylumCase)
+            .stream()
+            .map(StrategicCaseFlag::getDetails)
+            .flatMap(Collection::stream)
+            .filter(detail -> Objects.equals("Active", detail.getCaseFlagValue().getStatus()))
+            .toList();
+
+        List<CaseFlagDetail> languageFlags = new ArrayList<>();
+        List<CaseFlagDetail> reasonableAdjustmentsFlags = new ArrayList<>();
+
+        caseFlags
+            .forEach(flagDetail -> {
+                if (isLanguageCaseFlag(flagDetail)) {
+                    languageFlags.add(flagDetail);
+                } else if (isReasonableAdjustmentFlag(flagDetail)) {
+                    reasonableAdjustmentsFlags.add(flagDetail);
+                }
+            });
+
+        // get the languageCaseFlags sorted to have the ones with code above the ones without
+        List<CaseFlagDetail> sortedLanguageFlags = languageFlags.stream()
+            .sorted(Comparator.comparing(
+                detail -> detail.getCaseFlagValue().getSubTypeKey(),
+                Comparator.nullsLast(Comparator.naturalOrder())))
+            .toList();
+
+        // extract one flag with code to populate "interpreterLanguage" and group all the others with code
```

Comment:
For readability would it be better to extract the logics for fetching interpreterLanguageFlag to a method instead?

Reply:
I thought about it, but the logic isn't just for fetching interpreterLanguageFlag, but for fetch all language-related flags and identify one that is eligible to be used in `interpreterLanguage` and keep the rest to populate `otherReasonableAdjustmentDetails`, so the result of this logic would be two lists.